src/include/slab_containers.h

```diff
@@ -0,0 +1,376 @@
+// on the access pattern (and container type). The worst-case  memory
+// expansion is 'slabSize-1' times an STL container (approximately).
+// However it's likely to be very hard to create a worst-case memory
+// pattern (unless you're trying ;-).
+// This also leads toward keeping the slabSizes modest -- unless you
+// have special knowledge  of the access pattern OR the container itself
+// is expected to be ephemeral.
+//
+// Another disdvantage of slabs is that you can't move them from one
+// container to another. This causes certain operations that used to
+// have O(1) cost to have O(N) cost.
+//
+//    list::splice  O(N)
+//    list::swap    O(2N)
+//    vector::swap is O(2*stackSize)
+//
+// Not all of the stl::xxxx.swap operations are currently implemented.
+// They are pretty much discouraged for slab_containers anyways.
+//
+// In order to reduce malloc/free traffic further, each container has
+// within it one slab that is part of the container itself (logically
+// pre-allocated, but it's not a separate malloc call -- its part of
+// the container itself). The size of this slab is set by the template
+// parameter 'StackSlabSizeV' which must be specified on every
+// declaration (no default).  Thus the first 'StackSlabSizeV' items in
+// this slab container result in no malloc/free invocation at the expense
+// of burning that memory whether it's used or not. This is ideal for
+// containers that are normally one or two nodes but have the ability
+// to grow on that rare occasion when it's needed.
+//
+// There is only one way to create more slabs: When no more free nodes
+// are available, a slab is allocated using the HeapSlabSizeV template
+// parameter.
+
+template<typename T, std::size_t StackSlabSizeV, std::size_t HeapSlabSizeV>
+class slab_allocator {
+  struct heap_slab_t;
+
+  // In contrast to an embedded slot, each heap slot has overhead
+  // of one pointer.
+  struct heap_slot_t {
+    heap_slab_t* owner;
+    std::aligned_storage_t<sizeof(T), alignof(T)> storage;
+  };
+
+  // In contrast to an embedded slab, each heap slab can be tracked
+  // using a freelist. Together with the need for removal in O(1),
+  // this results in using a doubly linked list.
+  struct heap_slab_t : public boost::intrusive::list_base_hook<> {
+    // Control data should be put at the end of struct if we would
+    // like to take care about types with extended alignment.
+    ceph::math::bitset<HeapSlabSizeV> free_map;
+    std::array<heap_slot_t, HeapSlabSizeV> slots;
+  };
+
+  using stack_slot_t = std::aligned_storage_t<sizeof(T), alignof(T)>;
+
+  struct stack_slab_t {
+    ceph::math::bitset<StackSlabSizeV> free_map;
+    std::array<stack_slot_t, StackSlabSizeV> storage;
+  };
+
+  // The pointer-to-tail of doubly linked list is useless for us
+  // but we really need the ability to do O(1) list management
+  // being given only the pointer slab.
+  boost::intrusive::list<heap_slab_t> heap_free_list;
```

Comment:
FIXME: this guy stores size internally. We don't really need `heap_free_list.size()` to be O(1), so paying the extra management costs and worsening our cache density is unnecessary overhead. Kill with `constant_time_size<false>`.